lib/features.js

```diff
@@ -0,0 +1,105 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const hash = require('sha1')
+
+module.exports = config => {
+  const lastAccessTimeUpdates = config.lastAccessTimeUpdates
+  const signinConfirmation = config.signinConfirmation
+
+  return {
+    /**
+     * Predicate that indicates whether updates to sessionToken.lastAccessTime
+     * are enabled for a given user, based on ther uid and email address.
+     *
+     * @param uid   Buffer or String
+     * @param email String
+     */
+    isLastAccessTimeEnabledForUser (uid, email) {
+      return lastAccessTimeUpdates.enabled && (
+        isSampledUser(lastAccessTimeUpdates.sampleRate, uid, 'lastAccessTimeUpdates') ||
+        new RegExp(lastAccessTimeUpdates.enabledEmailAddresses).test(email)
+      )
+    },
+
+    /**
+     * Predicate that indicates whether sign-in confirmation is enabled
+     * for a given user, based on ther uid and email address.
+     *
+     * @param uid   Buffer or String
+     * @param email String
+     */
+    isSigninConfirmationEnabledForUser (uid, email, request) {
+      if (! signinConfirmation.enabled) {
+        return false
+      }
+
+      // Always create unverified tokens if customs-server
+      // has said the request is suspicious.
+      if (request.app.isSuspiciousRequest) {
+        return true
+      }
+
+      // Or if the email address matches one of these regexes.
+      if (signinConfirmation.forceEmailRegex.some(regex => new RegExp(regex).test(email))) {
+        return true
+      }
+
+      // While we're testing this feature, there may be some funky
+      // edge-cases in device login flows that haven't been fully tested.
+      // Temporarily avoid them for regular users by checking the `context` flag,
+      // and create pre-verified sessions for unsupported clients.
+      // This check will go away in the final version of this feature.
+      const context = request.payload &&
+        request.payload.metricsContext &&
+        request.payload.metricsContext.context
+      if (signinConfirmation.supportedClients.indexOf(context) === -1) {
+        return false
+      }
+
+      // Check to see if user in roll-out cohort.
+      return isSampledUser(signinConfirmation.sample_rate, uid, 'signinConfirmation')
+    },
+
+    /**
+     * Predicate that indicates whether a user belongs to the sampled cohort,
+     * based on a sample rate, their uid and a key that identifies the feature.
+     *
+     * @param sampleRate Number in the range 0..1
+     * @param uid        Buffer or String
+     * @param key        String
+     */
+    isSampledUser: isSampledUser
+  }
+}
+
+const HASH_LENGTH = hash('').length
+const MAX_SAFE_HEX = Number.MAX_SAFE_INTEGER.toString(16)
+const MAX_SAFE_HEX_LENGTH = MAX_SAFE_HEX.length - MAX_SAFE_HEX.indexOf('f')
+const COHORT_DIVISOR = parseInt(Array(MAX_SAFE_HEX_LENGTH).fill('f').join(''), 16)
```

Comment:
I thought it better to derive these constants from first principles rather than use magic numbers with explanatory comments. It means there is a bit more to read but it also means that they're robust in the event of (admittedly unlikely) changes to the hashing algorithm or the value of `Number.MAX_SAFE_INTEGER`.